### DISCUSSION TITLE
TusUploady Filename and Type

### DISCUSSION THREAD

**Discussion by chriskolenko** [rpldy/react-uploady#249] [Q&A] [answered]
Hi everyone,

We currently have gotten the TusUploady working with a custom DropZoneButton. The last thing we would like to do is add the filename and type as metadata. 

```
        <TusUploady
          destination={{ url: `${url}/api/assets/files` }}
          chunkSize={2142880}
        >
          <DropZoneButton extraProps={{ previewURL: previewURL }} />
        </TusUploady>
```

The DropZoneButton is based of the story book so we can support clicking the drop zone area and dropping files etc.

**Comment by yoavniran** [comment, marked as answer]
hey @chriskolenko 
from the Tus-Uploady [readme](https://github.com/rpldy/react-uploady/tree/master/packages/ui/tus-uploady): 

> Params prop that is set on the Destination or upload options is serialized (encoded according to Tus protocol) and sent as the value of the Upload-Metadata header.

You can set those explicitly from your destination object (on TusUploady or UploadButton) or you can set them dynamically using the [useRequestPreSend](https://github.com/rpldy/react-uploady/blob/master/packages/ui/uploady#userequestpresend-event-hook) hook.
The [Dynamic Parameters](https://github.com/rpldy/react-uploady/blob/master/guides/DynamicParameters.md) guide explains this concept in more detail.

**Reply by chriskolenko** [reply]
Hi @yoavniran,

Not having much success with useRequestPreSend. 

The file is empty inside the items array. So I can't get the filename or type

```
items : [{"id":"batch-1.item-1","batchId":"batch-1","state":"added","uploadStatus":0,"completed":0,"loaded":0,"aborted":false,"recycled":false,"previousBatch":null,"file":{}}]
```

**Reply by chriskolenko** [reply]
@yoavniran Ignore this. Sorry, I was using JSON.stringfy and the file isn't serializable.

This worked fine for my example:

```
  useRequestPreSend(({ items }) => {
    if (items.length == 0) {
      return {};
    }

    // we should only have 1 file for a TUS upload
    return {
      options: {
        params: {
          "file-name": items[0].file.name,
          "file-type": items[0].file.type,
        }
      }
    }
  });
```

Anytips on how to get the ID created from the server into a form outside of the TusUploady process?

**Reply by yoavniran** [reply]
thats great @chriskolenko 
I wasnt able to reproduce the issue obviously so thanks for clarifying.

as for the Tus ID, its currently inaccessible as its used only internally. 
I guess you could get it from the local storage item stored for the item being uploaded.

Could be an interesting enhancement. What's the use case though?

**Reply by chriskolenko** [reply]
I can get the ID from the location object which is good enough. 

```
  useItemFinalizeListener((item) => {
    if (!onChange) return

    const id = item?.uploadResponse?.location || ""
    onChange(id)
  });
```
